src/lib/visualization/abstract_visualizer.hpp

```diff
@@ -244,27 +244,43 @@ class AbstractVisualizer {
 
   std::string _wrap_label(const std::string& label) {
     if (label.length() <= MAX_LABEL_WIDTH) return label;
-
-    // Split by word so we don't break a line in the middle of a word
-    std::vector<std::string> label_words;
-    boost::split(label_words, label, boost::is_any_of(" "));
-
-    std::stringstream wrapped_label;
-    auto current_line_length = 0;
-
-    for (const auto& word : label_words) {
-      auto word_length = word.length() + 1;  // include whitespace
-
-      if (current_line_length + word_length > MAX_LABEL_WIDTH) {
-        wrapped_label << "\\n";
-        current_line_length = 0u;
+    std::stringstream label_stream;
+
+    // 1. Split label into lines
+    std::vector<std::string> lines;
+    boost::split(lines, label, boost::is_any_of("\n"));
+    const auto line_count = lines.size();
+    for (auto line_idx = size_t{0}; line_idx < line_count; ++line_idx) {
+      if (line_idx > 0) label_stream << '\n';
+      const auto& line = lines[line_idx];
+      if (line.length() <= MAX_LABEL_WIDTH) {
+        label_stream << line;
+        continue;
+      }
+      // 2. Split line into words, so we don't break a line in the middle of a word
+      std::vector<std::string> line_words;
+      boost::split(line_words, line, boost::is_any_of(" "));
+      size_t line_length = 0;
+      size_t word_idx = 0;
+      while (true) {
+        label_stream << line_words.at(word_idx);
+        line_length += line_words.at(word_idx).length();
+
+        // Exit on last word
+        if (word_idx == line_words.size() - 1) break;
+
+        line_length++;  // include whitespace
+        size_t next_line_length = line_length + line_words.at(++word_idx).length();
+        if (next_line_length < MAX_LABEL_WIDTH) {
```

Comment:
`<=`?

Reply:
Let's say you have a limit of 30. You have 19 chars already in place, add a space now. The next word is 10 chars long. You would currently no append the word, even though 19 + 1 + 10 is not over the max line limit.

You would immediately add a linebreak and start the next line.

Or am I wrong here?

Reply:
Oh. I did not notice that I already incremented

```cpp
line_length++;  // include whitespace
```

before the if. So you are right, thanks!